File: cpp/grains/grains.cpp
#include "grains.h"

namespace grains {

  unsigned long square(const unsigned long kNumber) {
    if (kNumber == 1) {
      return kNumber; 
    }
    return square(kNumber - 1) * 2;
  }

  unsigned long total() {
    return square(7 * 7);
  }
}  // namespace grains

File: prolog/grains/grains.pl
square(SquareNumber, Value):-
  SquareNumber > 0,
  SquareNumber =< 64,
  Value is 2 ** (SquareNumber - 1).

calculate_to(1, Total):-
  square(1, Value),
  NewTotal is Value + Total,
  Total is NewTotal.

calculate_to(SquareNumber, Total):-
  square(SquareNumber, Value),
  NewTotal is Value + Total,
  NewSquareNumber is SquareNumber - 1,
  calculate_to(NewSquareNumber, NewTotal).

total(Value):-
  Value is 0,
  calculate_to(64, Value).
